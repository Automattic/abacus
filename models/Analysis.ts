import parseISO from 'date-fns/fp/parseISO'

import { ApiData } from '@/api/ApiData'

/**
 * An analysis recommendation.
 */
class Recommendation {
  /**
   * @param endExperiment A boolean value indicating whether the experiment should end (based only on the raw data and
   *        ignoring any warnings).
   * @param chosenVariationId The ID of the variation that should be implemented if the experiment should end.
   *        This field is `null` if the experiment should continue running.
   * @param reason The reason for the recommendation. See `RecommendationReason`.
   * @param warnings An array of string warnings. See `RecommendationWarning`.
   */
  // istanbul ignore next (skip coverage for auto-generated constructor)
  constructor(
    public readonly endExperiment: boolean,
    public readonly chosenVariationId: number | null,
    public readonly reason: RecommendationReason,
    public readonly warnings: Array<RecommendationWarning>,
  ) {}

  /**
   * Create an instance from raw API data (parsed JSON).
   *
   * @param apiData Raw API data.
   */
  static fromApiData(apiData: ApiData): Recommendation {
    return {
      endExperiment: apiData.end_experiment,
      chosenVariationId: apiData.chosen_variation_id,
      reason: apiData.reason as RecommendationReason,
      warnings: apiData.warnings.map((warning: string) => warning as RecommendationWarning),
    }
  }
}

/**
 * Probabilistic estimate of a metric value.
 */
class MetricEstimate {
  /**
   * @param estimate Point estimate for the metric value.
   * @param bottom Bottom bound of the 95% credible interval.
   * @param top Top bound of the 95% credible interval.
   */
  // istanbul ignore next (skip coverage for auto-generated constructor)
  constructor(public readonly estimate: number, public readonly bottom: number, public readonly top: number) {}
}

/**
 * The strategy used for the analysis. One of the following:
 *  - `itt_pure`: Pure intention-to-treat -- all participants are analysed based on their initial variation assignment.
 *  - `mitt_no_spammers`: Modified intention-to-treat -- same as `itt_pure`, but excluding spammers that were
 *     flagged on the analysis datetime.
 *  - `mitt_no_crossovers`: Modified intention-to-treat -- same as `itt_pure`, but excluding participants that were
 *     assigned to multiple experiment variations before `analysis_datetime` (aka crossovers).
 *  - `mitt_no_spammers_no_crossovers`: Modified intention-to-treat -- same as `itt_pure`, but excluding both spammers
 *     and crossovers.
 *  - `pp_naive`: Naive per-protocol -- only participants that triggered one of the experiment's exposure events,
 *     excluding both spammers and crossovers.
 */
export enum AnalysisStrategy {
  IttPure = 'itt_pure',
  MittNoSpammers = 'mitt_no_spammers',
  MittNoCrossovers = 'mitt_no_crossovers',
  MittNoSpammersNoCrossovers = 'mitt_no_spammers_no_crossovers',
  PpNaive = 'pp_naive',
}

/**
 * The reason for the recommendation, describing the relationship between the credible interval (CI) and the region of
 * practical equivalence to zero (ROPE).
 */
export enum RecommendationReason {
  CiInRope = 'ci_in_rope',
  CiGreaterThanRope = 'ci_greater_than_rope',
  CiLessThanRope = 'ci_less_than_rope',
  CiRopePartlyOverlap = 'ci_rope_partly_overlap',
  RopeInCi = 'rope_in_ci',
}

/**
 * Warnings that may be added to a recommendation.
 *  - `short_period`: The experiment period is too short to draw a conclusion.
 *  - `long_period`: The experiment period is too long. It may be time to stop it.
 *  - `wide_ci`: The CI is too wide in comparison to the ROPE. If possible, it's best to collect more data.
 */
export enum RecommendationWarning {
  ShortPeriod = 'short_period',
  LongPeriod = 'long_period',
  WideCi = 'wide_ci',
}

/**
 * A single analysis instance. Typically, an experiment will have multiple analyses: One for each metric assignment,
 * analysis strategy, and analysis day.
 */
export class Analysis {
  /**
   * @param metricAssignmentId The metric assignment that this analysis is for.
   * @param analysisStrategy The strategy used for the analysis. See `AnalysisStrategy`.
   * @param participantStats Mapping from 'total', 'not_final', and 'variation_<id>' value to participant counts.
   * @param metricEstimates Mapping from 'diff' and 'variation_<id>' values to `MetricEstimate` objects.
   * @param recommendation Recommendation how to proceed based on the analysis. See `Recommendation`.
   * @param analysisDatetime Timestamp of the analysis.
   */
  // istanbul ignore next (skip coverage for auto-generated constructor)
  constructor(
    public readonly metricAssignmentId: number,
    public readonly analysisStrategy: AnalysisStrategy,
    public readonly participantStats: { [key: string]: number },
    public readonly metricEstimates: { [key: string]: MetricEstimate } | null,
    public readonly recommendation: Recommendation | null,
    public readonly analysisDatetime: Date,
  ) {}

  /**
   * Create an instance from raw API data (parsed JSON).
   *
   * @param apiData Raw API data.
   */
  static fromApiData(apiData: ApiData): Analysis {
    return {
      metricAssignmentId: apiData.metric_assignment_id,
      analysisStrategy: apiData.analysis_strategy as AnalysisStrategy,
      participantStats: apiData.participant_stats,
      // No need to use fromApiData() on the metric estimates since they're already in camelCase (single word keys).
      metricEstimates: apiData.metric_estimates,
      recommendation: apiData.recommendation ? Recommendation.fromApiData(apiData.recommendation) : null,
      analysisDatetime: parseISO(apiData.analysis_datetime),
    }
  }
}
