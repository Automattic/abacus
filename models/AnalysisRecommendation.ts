import { ApiData } from '@/api/ApiData'

export class AnalysisRecommendation {
  // /**
  //  * A boolean value indicating whether the experiment should end (based only on the raw data and ignoring any
  //  * warnings).
  //  */
  // endExperiment: boolean;
  //
  // /**
  //  * The ID of the variation that should be implemented if the experiment should end. This field is `null` if the
  //  * experiment should continue running.
  //  */
  // chosenVariationId: number | null;
  //
  // /**
  //  * The reason for the recommendation as a string that describes the relationship between the credible interval (CI)
  //  * and the region of practical equivalence to zero (ROPE).
  //  */
  // reason: AnalysisRecommendationReason;
  //
  // /**
  //  * An array of warnings (may be empty).
  //  */
  // warnings: Array<AnalysisRecommendationWarning>;

  constructor(
    public readonly endExperiment: boolean,
    public readonly chosenVariationId: number | null,
    public readonly reason: AnalysisRecommendationReason,
    public readonly warnings: Array<AnalysisRecommendationWarning>,
  ) {}

  static fromApiData(apiData: ApiData) {
    return new this(apiData.end_experiment, apiData.chosen_variation_id, apiData.reason, apiData.warnings)
  }
}

export enum AnalysisRecommendationReason {
  CiInRope = 'ci_in_rope',
  CiGreaterThanRope = 'ci_greater_than_rope',
  CiLessThanRope = 'ci_less_than_rope',
  CiRopePartlyOverlap = 'ci_rope_partly_overlap',
  RopeInCi = 'rope_in_ci',
}

export enum AnalysisRecommendationWarning {
  /**
   * The experiment period is too short to draw a conclusion.
   */
  ShortPeriod = 'short_period',

  /**
   * The experiment period is too long. It may be time to stop it.
   */
  LongPeriod = 'long_period',

  /**
   * The CI is too wide in comparison to the ROPE. If possible, it's best to collect more data.
   */
  WideCi = 'wide_ci',
}
