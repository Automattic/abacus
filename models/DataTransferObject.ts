// eslint-disable-next-line @typescript-eslint/no-explicit-any
type ExcludeMethods<T> = Pick<T, { [K in keyof T]: T[K] extends (_: any) => any ? never : K }[keyof T]>

/**
 * Base class for data transfer objects we get from the API.
 *
 * Used to work around TypeScript's lack of object initializers.
 * See: https://github.com/microsoft/TypeScript/issues/3895#issuecomment-488355369.
 */
export abstract class DataTransferObject<T> {
  public constructor(initializer: ExcludeMethods<T>) {
    Object.assign(this, initializer)
  }
}
