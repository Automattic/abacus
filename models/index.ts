export * from './Analysis'
export type { Event } from './Event'
export * from './ExperimentBare'
export * from './ExperimentFull'
export * from './MetricBare'
export * from './MetricFull'
export * from './MetricAssignment'
export * from './MetricRevenueParams'
export * from './Platform'
export * from './Segment'
export type { SegmentAssignment } from './SegmentAssignment'
export * from './Status'
export * from './Variation'
