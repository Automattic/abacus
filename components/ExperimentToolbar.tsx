import Button from '@material-ui/core/Button'
import Icon from '@material-ui/core/Icon'
import { createStyles, makeStyles, Theme } from '@material-ui/core/styles'
import Toolbar from '@material-ui/core/Toolbar'
import clsx from 'clsx'
import React from 'react'

import { ExperimentFull, Status } from '@/models'

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    disableButton: {
      color: theme.palette.error.main,
    },
    marginLeft: {
      marginLeft: theme.spacing(2),
    },
  }),
)

export type ExperimentToolbarMode = 'conclude' | 'disable' | 'edit' | 'view'

/* istanbul ignore next; not concerned whether this is called or not */
const noOp = () => undefined

/**
 * Renders a toolbar for an experiment.
 */
export default function ExperimentToolbar({
  className,
  experiment,
  mode,
  onCancel = noOp,
  onConclude = noOp,
  onDisable = noOp,
  onEdit = noOp,
  onSave = noOp,
}: {
  className?: string
  experiment: ExperimentFull
  mode: ExperimentToolbarMode
  onCancel?: () => void
  onDisable?: () => void
  onConclude?: () => void
  onEdit?: () => void
  onSave?: () => void
}) {
  const classes = useStyles()
  const { status } = experiment
  const hasConclusionData = experiment.hasConclusionData()
  const concludable = (status === Status.Completed || status === Status.Disabled) && !hasConclusionData
  const editable =
    status === Status.Staging ||
    status === Status.Running ||
    ((status === Status.Completed || status === Status.Disabled) && hasConclusionData)

  return (
    <Toolbar className={className} disableGutters>
      {(mode === 'disable' || mode === 'view') && status !== Status.Disabled && (
        <Button
          className={clsx(classes.disableButton, classes.marginLeft)}
          disabled={mode !== 'view'}
          onClick={() => onDisable()}
          variant='outlined'
        >
          Disable
        </Button>
      )}
      {editable &&
        (mode !== 'edit' ? (
          <Button
            className={classes.marginLeft}
            disabled={mode !== 'view'}
            onClick={() => onEdit()}
            startIcon={<Icon>edit</Icon>}
            variant='outlined'
          >
            Edit
          </Button>
        ) : (
          <>
            <Button className={classes.marginLeft} onClick={() => onCancel()} variant='outlined'>
              Cancel
            </Button>
            <Button className={classes.marginLeft} color='primary' onClick={() => onSave()} variant='contained'>
              Update Details
            </Button>
          </>
        ))}
      {concludable &&
        (mode !== 'conclude' ? (
          <Button
            className={classes.marginLeft}
            color='primary'
            disabled={mode !== 'view'}
            onClick={() => onConclude()}
            startIcon={<Icon>add_circle_outline</Icon>}
            variant='contained'
          >
            Add Conclusions
          </Button>
        ) : (
          <>
            <Button className={classes.marginLeft} onClick={() => onCancel()} variant='outlined'>
              Cancel
            </Button>
            <Button className={classes.marginLeft} color='primary' onClick={() => onSave()} variant='contained'>
              Save Conclusions
            </Button>
          </>
        ))}
    </Toolbar>
  )
}
