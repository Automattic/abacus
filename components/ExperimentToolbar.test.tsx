import { fireEvent, render } from '@testing-library/react'
import React from 'react'

import Fixtures from '@/helpers/fixtures'
import { Status } from '@/models'

import ExperimentToolbar from './ExperimentToolbar'

// |                     state                        |
// | exp status: | conclusions: | section: | mode:    | valid:
// | ----------- | ------------ | -------- | -------- | ------
// | complete    | no           | details  | conclude | yes
// | disabled    | no           | details  | conclude | yes
// | complete    | no           | details  | disable  | yes
// | complete    | yes          | details  | disable  | yes
// | running     | n/a          | details  | disable  | yes
// | staging     | n/a          | details  | disable  | yes
// | running     | n/a          | details  | edit     | yes
// | staging     | n/a          | details  | edit     | yes
// | complete    | no           | details  | view     | yes
// | complete    | yes          | details  | view     | yes
// | disabled    | no           | details  | view     | yes
// | disabled    | yes          | details  | view     | yes
// | running     | n/a          | details  | view     | yes
// | staging     | n/a          | details  | view     | yes
// | n/a         | n/a          | results  | view     | yes
// | n/a         | n/a          | snippets | view     | yes

test('with a complete experiment, no conclusions, on details section, and in conclude mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull()
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='conclude' section='details' />)

  // Expect "Cancel" and "Save Conclusions" buttons.
  expect(container).toMatchSnapshot()
})

test('with a disabled experiment, no conclusions, on details section, and in conclude mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull({
    status: Status.Disabled,
  })
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='conclude' section='details' />)

  // Expect "Cancel" and "Save Conclusions" buttons.
  expect(container).toMatchSnapshot()
})

test('with a complete experiment, no conclusions, on details section, and in disable mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull()
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='disable' section='details' />)

  // Expect disabled "Disable" and disabled "Add Conclusions" buttons.
  expect(container).toMatchSnapshot()
})

test('with a complete experiment, with conclusion, on details section, and in disable mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull({
    endReason: 'Ran its course.',
  })
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='disable' section='details' />)

  // Expect disabled "Disable" and disabled "Edit" buttons.
  expect(container).toMatchSnapshot()
})

test('with a running experiment, on details section, and in disable mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull({
    status: Status.Running,
  })
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='disable' section='details' />)

  // Expect disabled "Disable" and disabled "Edit" buttons.
  expect(container).toMatchSnapshot()
})

test('with a staging experiment, on details section, and in disable mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull({
    status: Status.Staging,
  })
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='disable' section='details' />)

  // Expect disabled "Disable" and disabled "Edit" buttons.
  expect(container).toMatchSnapshot()
})

test('with a running experiment, on details section, and in edit mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull({
    status: Status.Running,
  })
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='edit' section='details' />)

  // Expect "Cancel" and "Update Details" buttons.
  expect(container).toMatchSnapshot()
})

test('with a staging experiment, on details section, and in edit mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull({
    status: Status.Staging,
  })
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='edit' section='details' />)

  // Expect "Cancel" and "Update Details" buttons.
  expect(container).toMatchSnapshot()
})

test('with a complete experiment, no conclusions, on details section, and in view mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull()
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='view' section='details' />)

  // Expect "Disable" and "Add Conclusions" buttons.
  expect(container).toMatchSnapshot()
})

test('with a complete experiment, with conclusions, on details section, and in view mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull({
    endReason: 'Ran its course.',
  })
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='view' section='details' />)

  // Expect "Disable" and "Edit" buttons.
  expect(container).toMatchSnapshot()
})

test('with a disabled experiment, no conclusions, on details section, and in view mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull({
    status: Status.Disabled,
  })
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='view' section='details' />)

  // Expect "Add Conclusions" button.
  expect(container).toMatchSnapshot()
})

test('with a disabled experiment, with conclusion, on details section, and in view mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull({
    status: Status.Disabled,
    endReason: 'Ran its course.',
  })
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='view' section='details' />)

  // Expect "Edit" button.
  expect(container).toMatchSnapshot()
})

test('with a running experiment, on details section, and in view mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull({
    status: Status.Running,
  })
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='view' section='details' />)

  // Expect "Disable" and "Edit" buttons.
  expect(container).toMatchSnapshot()
})

test('with a staging experiment, on details section, and in view mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull({
    status: Status.Staging,
  })
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='view' section='details' />)

  // Expect "Disable" and "Edit" buttons.
  expect(container).toMatchSnapshot()
})

test('on results section and in view mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull()
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='view' section='results' />)

  // Expect no toolbar buttons.
  expect(container).toMatchSnapshot()
})

test('on snippets section and in view mode will render expected buttons', () => {
  const experiment = Fixtures.createExperimentFull()
  const { container } = render(<ExperimentToolbar experiment={experiment} mode='view' section='snippets' />)

  // Expect no toolbar buttons.
  expect(container).toMatchSnapshot()
})

test('clicking "Disable" button calls `onDisabled` callback', () => {
  const handleDisable = jest.fn()
  const experiment = Fixtures.createExperimentFull({ status: Status.Staging })
  const { getByText } = render(
    <ExperimentToolbar experiment={experiment} mode='view' onDisable={handleDisable} section='details' />,
  )

  const disableElmt = getByText('Disable')

  fireEvent.click(disableElmt)

  expect(handleDisable).toHaveBeenCalledTimes(1)
})

test('clicking "Edit" button calls `onEdit` callback', () => {
  const handleEdit = jest.fn()
  const experiment = Fixtures.createExperimentFull({ status: Status.Staging })
  const { getByText } = render(
    <ExperimentToolbar experiment={experiment} mode='view' onEdit={handleEdit} section='details' />,
  )

  const editElmt = getByText('Edit')

  fireEvent.click(editElmt)

  expect(handleEdit).toHaveBeenCalledTimes(1)
})

test('clicking "Cancel" button calls `onCancel` callback', () => {
  const handleCancel = jest.fn()
  const experiment = Fixtures.createExperimentFull({ status: Status.Staging })
  const { getByText } = render(
    <ExperimentToolbar experiment={experiment} mode='edit' onCancel={handleCancel} section='details' />,
  )

  const cancelElmt = getByText('Cancel')

  fireEvent.click(cancelElmt)

  expect(handleCancel).toHaveBeenCalledTimes(1)
})

test('clicking "Update Details" button calls `onSave` callback', () => {
  const handleSave = jest.fn()
  const experiment = Fixtures.createExperimentFull({ status: Status.Staging })
  const { getByText } = render(
    <ExperimentToolbar experiment={experiment} mode='edit' onSave={handleSave} section='details' />,
  )

  const updateDetailsElmt = getByText('Update Details')

  fireEvent.click(updateDetailsElmt)

  expect(handleSave).toHaveBeenCalledTimes(1)
})

test('clicking "Add Conclusions" button calls `onConclude` callback', () => {
  const handleConclude = jest.fn()
  const experiment = Fixtures.createExperimentFull()
  const { getByText } = render(
    <ExperimentToolbar experiment={experiment} mode='view' onConclude={handleConclude} section='details' />,
  )

  const addConclusionsElmt = getByText('Add Conclusions')

  fireEvent.click(addConclusionsElmt)

  expect(handleConclude).toHaveBeenCalledTimes(1)
})

test('clicking "Cancel" button calls `onCancel` callback', () => {
  const handleCancel = jest.fn()
  const experiment = Fixtures.createExperimentFull()
  const { getByText } = render(
    <ExperimentToolbar experiment={experiment} mode='conclude' onCancel={handleCancel} section='details' />,
  )

  const cancelElmt = getByText('Cancel')

  fireEvent.click(cancelElmt)

  expect(handleCancel).toHaveBeenCalledTimes(1)
})

test('clicking "Save Conclusions" button calls `onSave` callback', () => {
  const handleSave = jest.fn()
  const experiment = Fixtures.createExperimentFull()
  const { getByText } = render(
    <ExperimentToolbar experiment={experiment} mode='conclude' onSave={handleSave} section='details' />,
  )

  const saveConclusionsElmt = getByText('Save Conclusions')

  fireEvent.click(saveConclusionsElmt)

  expect(handleSave).toHaveBeenCalledTimes(1)
})
