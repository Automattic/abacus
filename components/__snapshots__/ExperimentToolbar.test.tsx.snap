// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with a complete experiment, no conclusions, and in conclude mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-marginLeft-2"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Cancel
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-marginLeft-2 MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Save Conclusions
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`with a complete experiment, no conclusions, and in disable mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-disableButton-5 makeStyles-marginLeft-6 Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Disable
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-marginLeft-6 MuiButton-containedPrimary Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <span
            aria-hidden="true"
            class="material-icons MuiIcon-root"
          >
            add_circle_outline
          </span>
        </span>
        Add Conclusions
      </span>
    </button>
  </div>
</div>
`;

exports[`with a complete experiment, no conclusions, and in view mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-disableButton-17 makeStyles-marginLeft-18"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Disable
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-marginLeft-18 MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <span
            aria-hidden="true"
            class="material-icons MuiIcon-root"
          >
            add_circle_outline
          </span>
        </span>
        Add Conclusions
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`with a complete experiment, with conclusion, and in disable mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-disableButton-7 makeStyles-marginLeft-8 Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Disable
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-marginLeft-8 Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <span
            aria-hidden="true"
            class="material-icons MuiIcon-root"
          >
            edit
          </span>
        </span>
        Edit
      </span>
    </button>
  </div>
</div>
`;

exports[`with a complete experiment, with conclusions, and in view mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-disableButton-19 makeStyles-marginLeft-20"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Disable
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-marginLeft-20"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <span
            aria-hidden="true"
            class="material-icons MuiIcon-root"
          >
            edit
          </span>
        </span>
        Edit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`with a disabled experiment, no conclusions, and in conclude mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-marginLeft-4"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Cancel
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-marginLeft-4 MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Save Conclusions
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`with a disabled experiment, no conclusions, and in view mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-marginLeft-22 MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <span
            aria-hidden="true"
            class="material-icons MuiIcon-root"
          >
            add_circle_outline
          </span>
        </span>
        Add Conclusions
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`with a disabled experiment, with conclusion, and in view mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-marginLeft-24"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <span
            aria-hidden="true"
            class="material-icons MuiIcon-root"
          >
            edit
          </span>
        </span>
        Edit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`with a running experiment and in disable mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-disableButton-9 makeStyles-marginLeft-10 Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Disable
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-marginLeft-10 Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <span
            aria-hidden="true"
            class="material-icons MuiIcon-root"
          >
            edit
          </span>
        </span>
        Edit
      </span>
    </button>
  </div>
</div>
`;

exports[`with a running experiment and in edit mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-marginLeft-14"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Cancel
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-marginLeft-14 MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Update Details
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`with a running experiment and in view mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-disableButton-25 makeStyles-marginLeft-26"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Disable
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-marginLeft-26"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <span
            aria-hidden="true"
            class="material-icons MuiIcon-root"
          >
            edit
          </span>
        </span>
        Edit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`with a staging experiment and in disable mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-disableButton-11 makeStyles-marginLeft-12 Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Disable
      </span>
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-marginLeft-12 Mui-disabled Mui-disabled"
      disabled=""
      tabindex="-1"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <span
            aria-hidden="true"
            class="material-icons MuiIcon-root"
          >
            edit
          </span>
        </span>
        Edit
      </span>
    </button>
  </div>
</div>
`;

exports[`with a staging experiment and in edit mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-marginLeft-16"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Cancel
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-marginLeft-16 MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Update Details
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;

exports[`with a staging experiment and in view mode will render expected buttons 1`] = `
<div>
  <div
    class="MuiToolbar-root MuiToolbar-regular"
  >
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-disableButton-27 makeStyles-marginLeft-28"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Disable
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-marginLeft-28"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium"
        >
          <span
            aria-hidden="true"
            class="material-icons MuiIcon-root"
          >
            edit
          </span>
        </span>
        Edit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</div>
`;
